src/directives/badge/badge.directive.ts

```diff
@@ -0,0 +1,204 @@
+import {
+    Directive,
+    ElementRef,
+    HostBinding,
+    Input,
+    OnChanges,
+    OnDestroy,
+    Renderer2,
+    SimpleChanges,
+} from '@angular/core';
+import { ColorService } from '../../services/color/color.service';
+import { ThemeColor } from '../../services/color/theme-color';
+import { ContrastService } from '../accessibility/contrast-ratio/contrast.service';
+
+export type BadgeVerticalPosition = 'above' | 'below';
+
+export type BadgeHorizontalPosition = 'before' | 'after';
+
+export type BadgeSize = 'small' | 'medium' | 'large';
+
+@Directive({
+    selector: '[uxBadge]',
+    exportAs: 'ux-badge',
+    host: {
+        class: 'ux-badge-container',
+        '[class.ux-badge-small]': 'badgeSize === "small"',
+        '[class.ux-badge-medium]': 'badgeSize === "medium"',
+        '[class.ux-badge-large]': 'badgeSize === "large"',
+        '[class.ux-badge-above]': 'badgeVerticalPosition === "above"',
+        '[class.ux-badge-below]': 'badgeVerticalPosition === "below"',
+        '[class.ux-badge-after]': 'badgeHorizontalPosition === "after"',
+        '[class.ux-badge-before]': 'badgeHorizontalPosition === "before"',
+    },
+})
+export class BadgeDirective implements OnChanges, OnDestroy /*, AfterContentChecked*/ {
+    private readonly _className = 'ux-badge';
+    private readonly _darkColor: ThemeColor = ThemeColor.parse('#000');
+    private readonly _lightColor: ThemeColor = ThemeColor.parse('#FFF');
+    private _badgeElement: HTMLElement | undefined;
+    private _isNumber: boolean;
+    private _truncatedText: boolean = false;
+
+    /**
+     * Directive parameter that sets the content of the badge
+     */
+    private _badgeDisplayContent: string;
+
+    @Input('uxBadge')
+    get badgeContent(): string {
+        return this._badgeContent;
+    }
+    set badgeContent(s: string) {
+        const subject = s.toString().trim();
+        this._isNumber = /^\d+$/.test(subject);
+        this._badgeContent = subject;
+    }
+    private _badgeContent: string;
+
+    /**
+     * Define the badge background color
+     */
+    private readonly _defaultBackgroundColor: ThemeColor = this._darkColor;
+
+    @Input()
+    get badgeColor(): string {
+        return this._badgeColor.toHex();
+    }
+    set badgeColor(color: string) {
+        this._badgeColor = ThemeColor.parse(this._colorService.resolve(color));
+    }
+    private _badgeColor: ThemeColor = this._defaultBackgroundColor;
+
+    /**
+     * Aria description value for accessibility devices
+     */
+    @Input() badgeAriaDescription: string;
```

Comment:
If we look a the aria-described by docs: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-describedby_attribute we can see that the value added to the element should refer to the id of the element used to describe the object.

If we look at what material does, it also adds an id to the badge element aria-describedby property:
![image](https://user-images.githubusercontent.com/20795331/78885349-56c7d600-7a54-11ea-91f7-be37702d01c3.png)

And then has the description in another element:
![image](https://user-images.githubusercontent.com/20795331/78885397-6c3d0000-7a54-11ea-8a7a-c761d5f898cd.png)

I'm guessing it is using https://material.angular.io/cdk/a11y/api#AriaDescriber however this is only supported by CDK v8+ and we are required to to support v7 for now, so perhaps we could create a simple service to do this ourselves (private and not documented) that has the same API surface as the CDK one (although only implement the functions we need) and then we can swap out for the CDK one when we drop v7. @donna-marijne thoughts?

Reply:
Hmm, I misunderstood this originally. Why would adding a hidden element for aria-describedby be better than aria-description? It'll be a string with no markup either way right?

Reply:
Yeah this confused me a bit when I noticed what they were doing in material. I'm guessing that they do it this way because a badge could be applied to an icon which can be aria-hidden? I'm not sure whether this is something we need to consider, but that's my guess as to why they do it this way? It's the only thing I can think of based on their docs.

If we don't think this is necessary then we might be able to get rid of this property entirely? As an aria label could just be added to the element that the button is attached to?

Reply:
That could be it, yes. But I think if the icon has meaning to a reader in combination with its badge then it probably have a tooltip attached with a proper description like "134 unread messages." And I agree that if it's a button or something, it can be given an aria-label or description which incorporates the information presented by the badge, e.g. "Show your messages. You have 134 unread messages."

Reply:
This is now removed